File: DataStructures/src/main/java/com/zhangdd/stack/ArrayStack.java
package com.zhangdd.stack;

/**
 * @author zhangdd on 2021/10/8
 */
public class ArrayStack {

    private final int maxSize;
    private final int[] stack;
    private int top = -1;

    public ArrayStack(int maxSize) {
        this.maxSize = maxSize;
        stack = new int[maxSize];
    }

    private boolean isFull() {
        return top == maxSize - 1;
    }

    private boolean isEmpty() {
        return top == -1;
    }

    public void push(int value) {
        if (isFull()) {
            System.out.println("当前栈已经满了");
            return;
        }
        top++;
        stack[top] = value;
    }

    public int pop() {
        if (isEmpty()) {
            throw new RuntimeException("当前栈空");
        }
        int value = stack[top];
        top--;
        return value;
    }

    public void list() {
        if (isEmpty()) {
            throw new RuntimeException("当前栈空");
        }
        for (int i = top; i >= 0; i--) {
            System.out.printf("stack[%d]=%d\n", i, stack[i]);
        }
    }
}

File: python/DataStructures/stack/ArrayStack.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/1/14
使用数组实现栈
"""


class ArrayStack:
    def __init__(self, max_size):
        self.top = -1
        self.stack = [None] * max_size
        self.max_size = max_size

    def pop(self):
        if self.top <= -1:
            print('当前栈为空')
            return
        value = self.stack[self.top]
        self.top -= 1
        return value

    def push(self, data):
        if self.top + 1 >= self.max_size:
            print('当前栈已满')
            return
        self.top += 1
        self.stack[self.top] = data

    def show(self):
        for i in range(self.top + 1):
            print(self.stack[i])


array_stack = ArrayStack(5)

print(len(array_stack.stack))
print('-' * 50)
array_stack.push(1)
array_stack.push(2)
array_stack.push(3)
array_stack.push(4)
array_stack.push(5)
array_stack.show()

array_stack.push(6)

print(array_stack.pop())
print(array_stack.pop())
array_stack.push(7)
print(array_stack.pop())
print(array_stack.pop())
print(array_stack.pop())
print(array_stack.pop())
print(array_stack.pop())
print(array_stack.pop())
